extras/kms/repository_root_key_version_test.go

```diff
@@ -523,3 +526,326 @@ func TestRepository_ListRootKeyVersions(t *testing.T) {
+		},
+		{
+			name:            "randReader-error",
+			writer:          rw,
+			repo:            testRepo,
+			rootWrapper:     rootWrapper,
+			rootKeyId:       rootKeyId,
+			opt:             []Option{WithRandomReader(newMockRandReader(1, "bad-reader"))},
+			wantErr:         true,
+			wantErrContains: "bad-reader",
+		},
+		{
+			name:            "Encrypt-error",
+			writer:          rw,
+			repo:            testRepo,
+			rootWrapper:     &mockTestWrapper{err: errors.New("Encrypt-error"), encryptError: true},
+			rootKeyId:       rootKeyId,
+			wantErr:         true,
+			wantErrContains: "unable to encrypt new root key version",
+		},
+		{
+			name:            "create-error",
+			writer:          &dbw.RW{},
+			repo:            testRepo,
+			rootWrapper:     rootWrapper,
+			rootKeyId:       rootKeyId,
+			wantErr:         true,
+			wantErrContains: "unable to create root key version",
+		},
+		{
+			name:        "success",
+			writer:      rw,
+			repo:        testRepo,
+			rootWrapper: rootWrapper,
+			rootKeyId:   rootKeyId,
+		},
+	}
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			assert, require := assert.New(t), require.New(t)
+
+			var currentRootKeyVersions []*rootKeyVersion
+			var encryptedBlob *wrapping.BlobInfo
+			if !tc.wantErr {
+				currentRootKeyVersions, err = tc.repo.ListRootKeyVersions(testCtx, tc.rootWrapper, tc.rootKeyId)
+				require.NoError(err)
+				sort.Slice(currentRootKeyVersions, func(i, j int) bool {
+					return currentRootKeyVersions[i].PrivateId < currentRootKeyVersions[j].PrivateId
+				})
+				currentWrapper, _, err := testKms.loadRoot(testCtx, testScopeId)
+				require.NoError(err)
+				encryptedBlob, err = currentWrapper.Encrypt(testCtx, []byte(testPlainText))
+				require.NoError(err)
+			}
+
+			got, err := rotateRootKeyVersionTx(testCtx, tc.writer, tc.rootWrapper, tc.rootKeyId, tc.opt...)
+			if tc.wantErr {
+				require.Error(err)
+				if tc.wantErrIs != nil {
+					assert.ErrorIs(err, tc.wantErrIs)
+				}
+				if tc.wantErrContains != "" {
+					assert.Contains(err.Error(), tc.wantErrContains)
+				}
+				return
+			}
+			require.NoError(err)
+			assert.NotNil(got)
+
+			// ensure we can decrypt something using the rotated wrapper (does
+			// the previous key still work)
+			rotatedWrapper, _, err := testKms.loadRoot(testCtx, testScopeId)
+			require.NoError(err)
+			pt, err := rotatedWrapper.Decrypt(testCtx, encryptedBlob)
+			require.NoError(err)
+			assert.Equal(testPlainText, string(pt))
+
+			// make sure the rotated rkv wasn't in the orig set
+			for _, currRkv := range currentRootKeyVersions {
+				assert.NotEqual(got, currRkv)
+			}
+
+			newRootKeyVersions, err := tc.repo.ListRootKeyVersions(testCtx, tc.rootWrapper, tc.rootKeyId)
+			require.NoError(err)
+			sort.Slice(newRootKeyVersions, func(i, j int) bool {
+				return newRootKeyVersions[i].PrivateId < newRootKeyVersions[j].PrivateId
+			})
+			assert.Equal(len(newRootKeyVersions), len(currentRootKeyVersions)*2)
```

Comment:
I think one thing that might be good to add here is after rotation ensuring that encrypting a value uses the new key version (and maybe is not decryptable with any of the previous key versions).

Reply:
done and I added this to the data key version unit test as well.